### ISSUE TITLE
Slow concat with large list of dataframes

### ISSUE THREAD

**Issue by kevinausmh** [pola-rs/polars#1183]
#### Describe your feature request

The current implementation of the concat function seems to scale badly. Or at least it does on my machine.
Maybe another solution could be considered, if someone can reproduce this.

Polars version: 0.8.22
Python
MacOS

Setup:
```python
import polars as pl

def large_concat(df_list):
    while len(df_list) > 1:
        df_list = [pl.concat(df_list[i : i + 2]) for i in range(0, len(df_list), 2)]
    return df_list[0]

df_list = [pl.DataFrame({"a": [i, i], "b": [i + 1, i + 1]}) for i in range(100000)]
```

Test with Polars concat:
```python
df_concat = pl.concat(df_list)
```
Timing:
```
CPU times: user 4min 38s, sys: 645 ms, total: 4min 39s
Wall time: 4min 39s
```

Merge only two dataframes with concat and iterate in Python:
```python
df_large_concat = large_concat(df_list)
```
Timing:
```
CPU times: user 610 ms, sys: 23.6 ms, total: 634 ms
Wall time: 634 ms
```

**Comment by kevinausmh** [comment]
I just saw, that concat is implemented in Python.
Maybe a Python version thing. I'm on Python 3.8.10

```python
def _stack(dfs):
    df = dfs[0].clone()
    if len(dfs) > 1:
        df.vstack(dfs[1].clone(), in_place=True)
    return df


def concat(dfs, rechunk: bool = True) -> "pl.DataFrame":
    while len(dfs) > 1:
        dfs = [_stack(dfs[i : i + 1]) for i in range(0, len(dfs), 2)]
    df = dfs[0]
    if rechunk:
        df = df.rechunk()
    return df
```
With a larger list
```python
df_list = [pl.DataFrame({"a": [i, i], "b": [i + 1, i + 1]}) for i in range(10000000)]
```

```python
df = large_concat(df_list)
```
```
CPU times: user 1min 5s, sys: 20.9 s, total: 1min 26s
Wall time: 1min 30s
```

```python
df = concat(df_list)
```
```
CPU times: user 20.2 s, sys: 1.86 s, total: 22.1 s
Wall time: 22.2 s
```

**Comment by ritchie46** [comment]
Thanks @kevinausmh . This is fixed in #1186

**Comment by ritchie46** [comment]
fixed and released.
